server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyConnectionCounter.scala

```diff
@@ -0,0 +1,20 @@
+package sttp.tapir.server.netty.internal
+
+import io.netty.channel.ChannelHandler.Sharable
+import io.netty.channel.{ChannelHandlerContext, ChannelInboundHandlerAdapter}
+
+import java.util.concurrent.atomic.AtomicInteger
+
+@Sharable case class NettyConnectionCounter(maxConnections: Int) extends ChannelInboundHandlerAdapter {
+  private val connections = new AtomicInteger()
+
+  override def channelActive(ctx: ChannelHandlerContext): Unit = {
+    val counter = connections.incrementAndGet
+    if (counter <= maxConnections) super.channelActive(ctx) else ctx.close
```

Comment:
**suggestion:** Before closing, let's log a `logger.warning(s"Max connections exceeded: $maxConnections")`
You can create a logger with
```scala
import io.netty.util.internal.logging.InternalLoggerFactory
// ...

lazy val logger = InternalLoggerFactory.getInstance(CombinedChannelDuplexHandler.class)
```

Afterwards, please test this locally with a low number, like 2. You can have an endpoint which sleeps for a few seconds in order to generate exceeding max connections. Please verify if this counter works as expected in such a case and if the warning is indeed logged.